File: src/main/arraylist/python/ArrayList.py
# O mais próximo que existe de um array em Python são Tuplas, que ainda assim 
# tem suas diferenças que comentarei quando chegar a hora
class ArrayList:
    def __init__(self, initial_data=()):
        self.tupla = tuple(initial_data)  # Garante que é uma tupla

    def isEmpty(self):
        return len(self.tupla) == 0

    def outOfBounds(self, index):
        if index < 0 or index >= len(self.tupla):
            raise IndexError(f"Index {index} fora dos limites (tamanho: {len(self.tupla)})")

    def addLast(self, elemento):
        return self.tupla + (elemento,)

    def addFirst(self, elemento):
        return (elemento,) + self.tupla

    def addIndex(self, elemento, index):
        self.outOfBounds(index)  
        return self.tupla[:index] + (elemento,) + self.tupla[index:]

    def rmvLast(self):
        if self.isEmpty():
            raise AttributeError("A lista está vazia")
        return self.tupla[:-1]

    def rmvFirst(self):
        if self.isEmpty():
            raise AttributeError("A lista está vazia")
        return self.tupla[1:]

    def rmvIndex(self, index):
        if self.isEmpty():
            raise AttributeError("A lista está vazia")
        self.outOfBounds(index)
        return self.tupla[:index] + self.tupla[index+1:]

    def contains(self, elemento):
        return elemento in self.tupla  

    def getFirst(self):
        if self.isEmpty():
            raise AttributeError("A lista está vazia")
        return self.tupla[0]

    def getLast(self):
        if self.isEmpty():
            raise AttributeError("A lista está vazia")
        return self.tupla[-1]

    def get(self, index):
        self.outOfBounds(index)
        return self.tupla[index]

    def size(self):
        return len(self.tupla)

File: src/main/arraylist/cpp/ArrayList.cpp
#include "ArrayList.h"
#include <algorithm>  // std::copy

ArrayList::ArrayList(size_t capacidadeInicial)
    : data(new int[capacidadeInicial]), capacity(capacidadeInicial), size(0) {}

ArrayList::~ArrayList() {
    delete[] data;
}

bool ArrayList::isEmpty() const {
    return size == 0;
}

size_t ArrayList::getSize() const {
    return size;
}

bool ArrayList::contains(int value) const {
    for (size_t i = 0; i < size; ++i) {
        if (data[i] == value) return true;
    }
    return false;
}

void ArrayList::testeTamanho() {
    if (size < capacity) return;
    size_t novaCap = capacity * 2;
    int* novoData = new int[novaCap];
    std::copy(data, data + size, novoData);
    delete[] data;
    data = novoData;
    capacity = novaCap;
}

void ArrayList::shiftDireita(size_t index) {
    for (size_t i = size; i > index; --i) {
        data[i] = data[i - 1];
    }
}

void ArrayList::shiftEsquerda(size_t index) {
    for (size_t i = index; i < size - 1; ++i) {
        data[i] = data[i + 1];
    }
}

bool ArrayList::addLast(int value) {
    testeTamanho();
    data[size++] = value;
    return true;
}

bool ArrayList::addFirst(int value) {
    return addIndex(value, 0);
}

bool ArrayList::addIndex(int value, size_t index) {
    if (index > size) {
        throw std::out_of_range("Índice fora do intervalo");
    }
    testeTamanho();
    shiftDireita(index);
    data[index] = value;
    ++size;
    return true;
}

bool ArrayList::rmvLast() {
    if (isEmpty()) return false;
    --size;
    return true;
}

bool ArrayList::rmvFirst() {
    return rmvIndex(0);
}

bool ArrayList::rmvIndex(size_t index) {
    if (index >= size) {
        throw std::out_of_range("Índice fora do intervalo");
    }
    shiftEsquerda(index);
    --size;
    return true;
}

int ArrayList::get(size_t index) const {
    if (index >= size) {
        throw std::out_of_range("Índice fora do intervalo");
    }
    return data[index];
}


